static/js/util/coupon.js

```diff
@@ -0,0 +1,85 @@
+// @flow
+import {
+  COUPON_CONTENT_TYPE_PROGRAM,
+  COUPON_CONTENT_TYPE_COURSE,
+  COUPON_CONTENT_TYPE_COURSERUN,
+  COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT,
+  COUPON_AMOUNT_TYPE_FIXED_DISCOUNT,
+} from '../constants';
+import type {
+  Coupons,
+  Coupon,
+  CalculatedPrices,
+} from '../flow/couponTypes';
+import type {
+  CoursePrice,
+  CoursePrices,
+  Dashboard,
+} from '../flow/dashboardTypes';
+
+export const calculatePrices = (programs: Dashboard, prices: CoursePrices, coupons: Coupons): CalculatedPrices => {
+  let couponLookup: Map<number, Coupon> = new Map();
+  let priceLookup: Map<number, CoursePrice> = new Map();
+  for (const coupon of coupons) {
+    couponLookup.set(coupon.program_id, coupon);
+  }
+  for (const price of prices) {
+    priceLookup.set(price.program_id, price);
+  }
+
+  return programs.map(program => ({
+    id: program.id,
+    courses: program.courses.map(course => ({
+      id: course.id,
+      runs: course.runs.map(run => ({
+        id: run.id,
+        price: mockableCalculateRunPrice(
+          run.id, course.id, program.id, priceLookup.get(program.id), couponLookup.get(program.id)
+        ),
+      }))
+    }))
+  }));
```

Comment:
this is also something that strikes me as good to make a little more 'flat', three nested `.map`s is kinda hard to read. you could separate it out into three separate functions, an inner one that deals with `runs`, another that deals with `courses`, etc.